pkg/components/linkerd/component.go

```diff
@@ -0,0 +1,186 @@
+	diagnostics := hcl.Diagnostics{}
+
+	if configBody == nil {
+		return hcl.Diagnostics{}
+	}
+
+	d := gohcl.DecodeBody(*configBody, evalContext, c)
+	if d.HasErrors() {
+		return append(diagnostics, d...)
+	}
+
+	if c.ControllerReplicas < 1 {
+		return hcl.Diagnostics{
+			{
+				Severity: hcl.DiagError,
+				Summary:  "'controller_replicas' should be more than or equal to one",
+			},
+		}
+	}
+
+	return diagnostics
+}
+
+func (c *component) RenderManifests() (map[string]string, error) {
+	// linkerd2 is the name of the upstream chart.
+	helmChart, err := util.LoadChartFromAssets("/components/linkerd2")
+	if err != nil {
+		return nil, fmt.Errorf("load chart from assets: %w", err)
+	}
+
+	if err = c.generateCertificates(); err != nil {
+		return nil, err
+	}
+
+	values, err := internaltemplate.Render(chartValuesTmpl, c)
+	if err != nil {
+		return nil, fmt.Errorf("rendering values template failed: %w", err)
+	}
+
+	values, err = mergeValuesFiles(values, valuesHA)
+	if err != nil {
+		return nil, fmt.Errorf("merging values failed: %w", err)
+	}
+
+	// Generate YAML for the Linkerd deployment.
+	renderedFiles, err := util.RenderChart(helmChart, name, c.Metadata().Namespace.Name, values)
+	if err != nil {
+		return nil, fmt.Errorf("rendering chart failed: %w", err)
+	}
+
+	return renderedFiles, nil
+}
+
+func (c *component) Metadata() components.Metadata {
+	return components.Metadata{
+		Name: name,
+		Namespace: k8sutil.Namespace{
+			Name: "linkerd",
+			Annotations: map[string]string{
+				"linkerd.io/inject": "disabled",
+			},
+			Labels: map[string]string{
+				"linkerd.io/is-control-plane":          "true",
+				"config.linkerd.io/admission-webhooks": "disabled",
+				"linkerd.io/control-plane-ns":          "linkerd",
+			},
+		},
+		Helm: components.HelmMetadata{
+			Wait: true,
+		},
+	}
+}
+
+func indent(data string, indent int) string {
+	lines := strings.Split(data, "\n")
+
+	var gap string
+
+	// Calculate the gap/indent.
+	for i := 0; i < indent; i++ {
+		gap += " "
+	}
+
+	// For each line add the gap/indent.
+	for ind := range lines {
+		lines[ind] = gap + lines[ind]
+	}
+
+	// If the last line is empty then remove the indent from it.
+	if lines[len(lines)-1] == gap {
+		lines[len(lines)-1] = ""
+	}
+
+	return strings.Join(lines, "\n")
+}
+
+func mergeValuesFiles(dst, src string) (string, error) {
+	d, err := chartutil.ReadValues([]byte(dst))
+	if err != nil {
+		return "", fmt.Errorf("could not read values from destination: %w", err)
+	}
+
+	s, err := chartutil.ReadValues([]byte(src))
+	if err != nil {
+		return "", fmt.Errorf("could not read values from source: %w", err)
+	}
+
+	return chartutil.Values(chartutil.CoalesceTables(d, s)).YAML()
+}
+
+func (c *component) generateCertificates() error {
+	root, err := tls.GenerateRootCAWithDefaults(certCommonName)
+	if err != nil {
+		return fmt.Errorf("generating certificates: %w", err)
+	}
+
+	c.Cert = cert{
```

Comment:
IMO this function should return this struct, not still mutate the component :) `(c *component)` receiver can then be removed too.

Reply:
I would like to know the reasoning behind one vs other?

Reply:
This one is stateful, as it mutates the receiver. What I propose it stateless and is easier to test, as the mutation happens outside of the function.